### ISSUE TITLE
Questions about two-stage regression

### ISSUE THREAD

**Issue by dremaker52** [sfzhang15/RefineDet#6]
Hi, the ODM and ARM use the same features for anchor regression so I think it's the same as only using ODM to do regression. I can understand that in Faster R-CNN, the regression in RPN is useful because it changes the features for second stage regression. But in this work, the two-stage regression looks like two independent regressions and the results are merged at the final. Do you have some comments on why this two-stage regression is better than only use ODM for regression? Thanks.

**Comment by sfzhang15** [comment]
@dremaker52  Hi, 
1. The ODM and ARM do not use the same features for anchor regression. ARM use the features in the above part, while ODM use the features in the below part. The feature in ODM is generated by TCBs, which combine the ARM's feature and higher layers' feature to make the feature in ODM richer, it is vital for ODM to handle the more challenging tasks, i.e., predict more accurate object locations, sizes and multi-class labels.
2.  The two-stage regression is not two independent regressions. The ARM is aims to regresses from anchor (**A**) to gt (**G**), and its regression result is **R**. There is still some error between  **R** and **G**. The ODM is aims to regresses from **R** to **G**, and its regression result is **T**. Comparing to **R**, **T** will be more accurate.

**Comment by dremaker52** [comment]
Thank you very much for your reply. It's my mistake to think that ODM and ARM use same features for anchor regression. However,  I still have some doubts about using two-stage regression. When ODM performs anchor regression, it has no information about R and can't get benefit from the refined anchors. So T will be more accurate is due to the better features (combine higher layers' feature) not the first stage regression R? If that's the case, will combining R to T break down the accuracy of T? Thanks.

**Comment by sfzhang15** [comment]
@dremaker52 Hi, 
When ODM performs anchor regression, it has the location information about R, since ODM takes R as the starting point for regression. The ARM provides the ODM with a better starting point for regression, just like the RPN provides the Fast R-CNN with a better starting point for regression. R corresponds to the regression result of RPN and T corresponds to the regression result of Fast R-CNN. Although ODM does not have the RoIPooling operation to crop the feature of R to be used for the next regression ( this is how Fast R-CNN does), but ODM has better feature with richer receptive field, which can also be used for the next regression from a better starting point to get better location.

**Comment by dremaker52** [comment]
Thanks!

**Comment by jay2002** [comment]
Hi @sfzhang15, in my opinion ARM and ODM perform regression twice and the regressions are conducted in a cascade manner. The overall procedure is like fast rcnn + fpn (performing 2nd regression in different featuremaps). 
Why are you calling this method as 'Single-Shot' method?

**Comment by sfzhang15** [comment]
Hi @jay2002 , in my opinion, the RoIPooling operation needs to crop feature, which corresponds to anchor 'shot'.

**Comment by jay2002** [comment]
I got your point. thx!

**Comment by jwnsu** [comment]
name does not matter, as long as it's fast and accurate. The speed of RefineDet is on par with SSD, with better accuracy. Nice work!

Multi-scale testing is very slow though, so it's more to claim accuracy crown than practical use :-).

**Comment by longyin880815** [comment]
Thank you for interested in our work! Due to the shortage of computational resources, we only train RefineDet with two kinds of input size 320x320 and 512x512. The multi-scale testing is used to reduce the impact of input size for a fair comparison. :)

**Comment by dremaker52** [comment]
Hi @longyin880815, When I trained the network, it takes about 5.6sec/per iteration for batch size=2 and iter_size = 3 on 4 K80 GPUs for resnet101 and 512*512. Is the speed normal as it will take about 35 days to train a single model?

**Comment by sfzhang15** [comment]
@dremaker52 Hi, we use 4 M40 GPUs to train RefineDet512_ResNet101_COCO and it takes about 2.7sec/per iteration for batch size=20 (every GPU has 5 images). Notably, if every GPU card has less than 4 images, which will cause unstable results in the BN layer and hurt accuracy.

**Comment by dremaker52** [comment]
Thanks. I found the training loss is decreasing very slow and it may be due to the batch size of 2 is too small for BN layer.

**Comment by lengly** [comment]
Really interesting! I think your work is trying to combine two-shot methods and one-shot methods together :-)     two stage regression is certainly better than one stage regression

**Comment by sfzhang15** [comment]
@lengly Yes! We let one-shot methods try to simulate two-shot methods‘ framework, in order to make one-shot methods have several advantages of two-shot methods, such as two-stage classification, two-stage regression and two-stage feature.
